lib/Dialect/TritonNvidiaGPU/IR/Ops.cpp

```diff
@@ -160,6 +160,18 @@ void WaitBarrierOp::getEffects(
                        mlir::triton::gpu::SharedMemory::get());
 }
 
+// -- TensorDescToTMAPtrOp --
+LogicalResult TensorDescToTMAPtrOp::canonicalize(TensorDescToTMAPtrOp op,
+                                                 PatternRewriter &rewriter) {
+  // tensor_desc_to_tma_ptr(reinterpret_tensor_desc(ptr)) -> ptr
+  if (auto reinterpret =
+          op.getDesc().getDefiningOp<triton::ReinterpretTensorDescOp>()) {
+    rewriter.replaceOp(op, reinterpret.getRawDesc());
+    return success();
+  }
+  return failure();
+}
```

Comment:
Welp I tried this out and it revealed there's a bit of a bug here. `tensor_desc_to_tma_ptr` returns a pointer to global memory, but by-val TMA is represented as a pointer to the generic address space. 

On the one hand this is a real concern because the by-val descriptor lives in parameter space, not global memory so the pointer type is wrong. On the other hand we only use it the result for TMA operations which are inline assembly anyway so it makes no practical difference. 

I'll just leave it as a canonicalizer for now, but in future I should figure out how to return the correct type from `tensor_desc_to_tma_ptr`.

Reply:
ok that works, isn't there a type mistmatch still? I guess it doesn't show up because the user can take either memory space?

Reply:
That's right, the only users of this op are the `AsyncTMA` ops, neither of which actually care about the address space.